### PR TITLE
Add support to Promise.prototype.then for Symbol.species

### PR DESCRIPTION
**Pull request by boingoing**
Add support to Promise.prototype.then for Symbol.species

Fixes a couple of issues with Promise and how it handled @@species.

First, Promise.prototype.then did not correctly use the @@species property to construct the promise stored in the PromiseCapability. See NewPromiseCapability (https://tc39.github.io/ecma262/#sec-newpromisecapability).

Next, we were looking at the @@species property in Promise.all and Promise.race when we are supposed to ignore it. Those API should use the 'this' argument (Promise in the nominal case) to construct the object they return.

Also cleaned up some stale comments, updated a few small details (like length property of anonymous functions used by Promise machinery), and fixed a bug where the executor function passed to the Promise constructor via Promise.all or Promise.race could not be called twice even if the first call passed undefined for arguments. This should be allowed.

### CODE REVIEW

**Comment by boingoing** [comment]
@abchatra @suwc Care to take a look? This fixes #538

**Comment by boingoing** [comment]
@akroshg Updated this PR to address 6127686, want to take a quick look over it?

File: test/es6/ES6Promise.js [hunk 1]
```diff
@@ -91,10 +91,10 @@ var tests = [
     {
         name: "Promise constructor throwing behavior",
         body: function () {
-            assert.throws(function() { Promise.call(); }, TypeError, "Promise throws when not called as a new expression with no this parameter", "Promise: 'this' is not a Promise object");
-            assert.throws(function() { Promise.call(undefined); }, TypeError, "Promise throws when not called as a new expression if the this parameter is undefined", "Promise: 'this' is not a Promise object");
-            assert.throws(function() { Promise.call(null); }, TypeError, "Promise throws when not called as a new expression if the this parameter is null", "Promise: 'this' is not a Promise object");
-            assert.throws(function() { Promise.call({}); }, TypeError, "Promise throws when not called as a new expression if the this parameter is not a promise", "Promise: 'this' is not a Promise object");
+            assert.throws(function() { Promise.call(); }, TypeError, "Promise throws when not called as a new expression with no this parameter", "'new.target' is undefined");
```
Comment (akroshg) [inline test/es6/ES6Promise.js]: Nit : somehow I don't like the new error as well - it is not intuitive. shouldn't this be "Promise constructor to be called as 'new' or 'Promise constructor requires the promise object' ?

File: lib/Runtime/Library/JavascriptPromise.cpp [hunk 2]
```diff
@@ -167,11 +165,6 @@ namespace Js
             JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NeedObject, _u("Promise.all"));
         }
 
-        // 3. Let S be Get(C, @@species).
-        // 4. ReturnIfAbrupt(S).
-        // 5. If S is neither undefined nor null, let C be S.
-        constructor = JavascriptOperators::GetSpecies(RecyclableObject::FromVar(constructor), scriptContext);
-
         JavascriptLibrary* library = scriptContext->GetLibrary();
         Var undefinedVar = library->GetUndefined();
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptPromise.cpp]: Nit: you can get rid of this var as well ? it is just used once I think.

File: lib/Runtime/Library/JavascriptPromise.cpp [hunk 3]
```diff
@@ -815,9 +782,18 @@ namespace Js
...
-        Assert(args.Info.Count > 1);
+        Var arg;
+
+        if (args.Info.Count > 1 && args[1] != nullptr)
+        {
+            arg = args[1];
+        }
+        else
+        {
+            arg = function->GetScriptContext()->GetLibrary()->GetUndefined();
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptPromise.cpp]: Should we still assert here?

**Comment by akroshg** [comment]
Rest looks good to me.

File: test/es6/ES6Promise.js (hunk 1 above)
Comment (boingoing) [inline test/es6/ES6Promise.js]: The meaning of the message is "Promise cannot be called without the 'new' keyword" but spec uses the case where new.target is undefined to indicate this. I don't think there is a way for new.target to actually be undefined if we do have the new flag, though. Even Reflect.construct(Promise,[],undefined) throws. I'll change the error message here.

File: lib/Runtime/Library/JavascriptPromise.cpp (hunk 2 above)
Comment (boingoing) [inline lib/Runtime/Library/JavascriptPromise.cpp]: We used this var a few times before the refactor here. Nice catch.

File: lib/Runtime/Library/JavascriptPromise.cpp (hunk 3 above)
Comment (boingoing) [inline lib/Runtime/Library/JavascriptPromise.cpp]: Spec says the Thrower function should return a completion with the value equal to the argument. So, I think it's correct for this function to be called with no argument and return undefined. I'll add assert that the arg is not nullptr here.

**Pull request merged**